kuksa_databroker/databroker/src/glob.rs

```diff
@@ -51,3 +62,131 @@ pub fn to_regex(glob: &str) -> Result<Regex, Error> {
     let re = to_regex_string(glob);
     Regex::new(&re).map_err(|_err| Error::RegexError)
 }
+
+pub fn is_valid_pattern(input: &str) -> bool {
+    let re = regex::Regex::new(
+        r"(?x)
+        ^  # anchor at start (only match full paths)
+        # At least one subpath (consisting of either of three):
+        (?:
+            [A-Z][a-zA-Z0-9]*  # An alphanumeric name (first letter capitalized)
+            |
+            \*                 # An asterisk
+            |
+            \*\*               # A double asterisk
+        )
+        # which can be followed by ( separator + another subpath )
+        # repeated any number of times
+        (?:
+            \. # Separator, literal dot
+            (?:
+                [A-Z][a-zA-Z0-9]*  # Alphanumeric name (first letter capitalized)
+                |
+                \*                 # An asterisk
+                |
+                \*\*               # A double asterisk
+            )
+        )*
+        $  # anchor at end (to match only full paths)
+        ",
+    )
+    .unwrap();
```

Comment:
Tip. Instead of building the regex every time this function is called, it's better to put it in a static variable using [lazy_static!](https://docs.rs/lazy_static/latest/lazy_static/) which would only be initialized once.

E.g.
```rust
use lazy_static::lazy_static;

lazy_static! {
    static ref REGEX_VALID_PATTERN: regex::Regex = regex::Regex::new(
        r"(?x)

        ...

        ",
    )
    .unwrap();
}

```

Reply:
if _Lazy_ static is doing what I think it does, is an "unwrap()" ok here, according to your no  `unwrap()` rule? I understand, that for stuff  during initialisation it is fine (because well, either it works, or databroker does not start), but wouldn't the lazy here indicate this might be executed anytime later during runtime? Or we can force the issue by using it already once somewhere during init?

Reply:
> if Lazy static is doing what I think it does
It is

> is an "unwrap()" ok here
Good point! Generally, no. But as the input for the regex will _never_ change (at least not at runtime) we can guarantee<sup>1</sup> that it will always succeed.

Now that you point it out, it should probably be changed to
```diff
--    .unwrap();
++    .expect("regex compilation (of static pattern) should always succeed")
```
or something. We could also force [initialization](https://docs.rs/lazy_static/latest/lazy_static/fn.initialize.html) of the static at startup (instead of first use).

`1` Guaranteed because it's used in our unit tests.